test/runner/lib/integration/__init__.py

```diff
@@ -0,0 +1,214 @@
+    ApplicationError,
+    display,
+    make_dirs,
+)
+
+from lib.cache import (
+    CommonCache,
+)
+
+
+def generate_dependency_map(integration_targets):
+    """
+    :type integration_targets: list[IntegrationTarget]
+    :rtype: dict[str, set[IntegrationTarget]]
+    """
+    targets_dict = dict((target.name, target) for target in integration_targets)
+    target_dependencies = analyze_integration_target_dependencies(integration_targets)
+    dependency_map = {}
+
+    invalid_targets = set()
+
+    for dependency, dependents in target_dependencies.items():
+        dependency_target = targets_dict.get(dependency)
+
+        if not dependency_target:
+            invalid_targets.add(dependency)
+            continue
+
+        for dependent in dependents:
+            if dependent not in dependency_map:
+                dependency_map[dependent] = set()
+
+            dependency_map[dependent].add(dependency_target)
+
+    if invalid_targets:
+        raise ApplicationError('Non-existent target dependencies: %s' % ', '.join(sorted(invalid_targets)))
+
+    return dependency_map
+
+
+def get_files_needed(target_dependencies):
+    """
+    :type target_dependencies: list[IntegrationTarget]
+    :rtype: list[str]
+    """
+    files_needed = []
+
+    for target_dependency in target_dependencies:
+        files_needed += target_dependency.needs_file
+
+    files_needed = sorted(set(files_needed))
+
+    invalid_paths = [path for path in files_needed if not os.path.isfile(path)]
+
+    if invalid_paths:
+        raise ApplicationError('Invalid "needs/file/*" aliases:\n%s' % '\n'.join(invalid_paths))
+
+    return files_needed
+
+
+@contextlib.contextmanager
+def integration_test_environment(args, target, inventory_path):
+    """
+    :type args: IntegrationConfig
+    :type target: IntegrationTarget
+    :type inventory_path: str
+    """
+    vars_file = 'integration_config.yml'
+
+    if args.no_temp_workdir or 'no/temp_workdir/' in target.aliases:
+        display.warning('Disabling the temp work dir is a temporary debugging feature that may be removed in the future without notice.')
+
+        integration_dir = 'test/integration'
+        ansible_config = os.path.join(integration_dir, '%s.cfg' % args.command)
+
+        inventory_name = os.path.relpath(inventory_path, integration_dir)
+
+        if '/' in inventory_name:
+            inventory_name = inventory_path
+
+        yield IntegrationEnvironment(integration_dir, inventory_name, ansible_config, vars_file)
+        return
+
+    root_temp_dir = os.path.expanduser('~/.ansible/test/tmp')
+
+    prefix = '%s-' % target.name
+    suffix = u'-\u00c5\u00d1\u015a\u00cc\u03b2\u0141\u00c8'
+
+    if args.no_temp_unicode or 'no/temp_unicode/' in target.aliases:
+        display.warning('Disabling unicode in the temp work dir is a temporary debugging feature that may be removed in the future without notice.')
+        suffix = '-ansible'
+
+    if isinstance('', bytes):
```

Comment:
Why don't you just encode suffix like `u'-\u00c5\u00d1\u015a\u00cc\u03b2\u0141\u00c8'.encode('utf-8')` and for the no_temp_unicode override just do `suffix = b'-ansible'`? No need for this check which is arguably more confusing.

Reply:
The encoding should only be done on Python 2.x, so it can't be done inline.

The `isinstance('', bytes)` check could be replaced with a check on `sys.version_info[0]` instead.

Reply:
Ok that makes sense.